python/mxnet/_ctypes/ndarray.py

```diff
@@ -31,21 +31,24 @@
 
 class NDArrayBase(object):
     """Base data structure for ndarray"""
-    __slots__ = ["handle", "writable"]
+    __slots__ = ["handle", "writable", "dlpack"]
     # pylint: disable= no-member
 
-    def __init__(self, handle, writable=True):
+    def __init__(self, handle, writable=True, dlpack=None):
```

Comment:
dlpack should not be part of the member, the PyCapsule manages itself

Reply:
When calling `FromDLPack` function, NDArray need to manage the release of `NDArrayDLManager::handle` in `ndarray.cc`.
e.g.
```
a = mx.nd.array([1,2,3]) # Denote TBlob x to store the data
pack = a.to_dlpack_for_write()
b = mx.nd.from_dlpack(pack)
del a, pack
# a and PyCapsule pack has been released.
# b need to manage the release TBlob x.
```
NDArray doesn't have the deleter function, so I made dlpack as a member of NDArray.
When the reference count of dlpack in NDArray is zero, the TBlob will be released.
Is there any other way to keep the reference?

Reply:
A better way is to keep NDArray's shared_ptr inside the manager_ctx itself, you can take a look at TVM's NDArray to DLManagedTesnor impl

Reply:
NDArray in MXNet and TVM are different. NDArray in TVM has the function `IncRef` and `DecRef` to change the reference count, however that in MXNet uses `NDArray::ptr_` (std::shared_ptr<Chunk>) to manage the reference count. `NDArray::ptr_` is a private member of `NDArray`.
The PR is similar to the PyTorch DLPack implementation [Code](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/DLConvertor.cpp#L128)
I add a `NDArrayDLManager` to manage the reference count of NDArray. [Code](https://github.com/apache/incubator-mxnet/pull/12047/files#diff-67d2910557c85e38570cd7c4f940cdbdR315) line 315 in src/ndarray/ndarray.cc

Setting `dlpack` as the NDArray(Python Class) member is to avoid the release of NDArrayDLManager, e.g. When the original NDArray and the PyCapsule (DLPack) are release, the new NDArray (generated by from_dlpack) still exists.

Reply:
you can create a new NDArray() that copies the original NDArray(which increases refcount) and put that as a context

Reply:
I use NDArrayDLManager whose member is a copy of NDArray, which increases refcount.
The deleter of DLPack is to delete NDArrayDLManager, which decreases refcount.

If a new NDArray() created by DLPack, the deleter of DLPack will be called by the new NDArray. Similarly, If a PyTorch Tensor createed by DLPack (MXNet), the deleter will be called by the new PyTorch Tensor.

NDArray (CPP and Python class) doesn't have deleter API, so I set the DLPack as the member of NDArray (Python). When NDArray generated by DLPack releases, the DLPack object in the NDArray will release, then the deleter function in DLPack will be called.

Reply:
You just need to create new NDArray that copies the original NDArray, and create a Deleter function that deletes the NDArray in the context. There is no need for python NDArray to keep reference of the dlpack object. As the DLManagedTensor manages itself

Reply:
Actually, your implementation of ToDLPack is fine and the Manager class is OK. Since there is already the Manager class and the deleter will get called, you do not need to keep reference to the dlpack in the NDArray

Reply:
Sorry. I don't know which object manages the actual memory after from_dlpack is called.
In the following code, 
```
a = mx.nd.array([1,2,3]) # Denote TBlob x to store the data
pack = a.to_dlpack_for_write()
b = mx.nd.from_dlpack(pack)
del a, pack
```
The original NDArray `a` and dlpack are released. However, the Manager `NDArrayDLManager` should be kept.
The new NDArray `b` get the data pointer from DLPack.
Which object will call the deleter?

In PyTorch, the Tensor class has the deleter structrual parameter. [Code1](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/DLConvertor.cpp#L170) [Code2](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/templates/Type.h#L88)
When the refcount of Tensor is zero, the Tensor Object will call the deleter to release `ATenDLMTensor`.
MXNet doesn't have the explicit deleter structual parameter. 
So I set `dlpack` as the member of NDArray (Python) class.
When the refcount of NDArray is zero, the refcount of dlpack in NDArray is zero too.
The dlpack will be released then call the deleter to delete `NDArrayDLManager`.

Reply:
In your case, when a get deleted, b still holds a NDArrayDLManager, which is allocated by new, and that object still hold NDArray(which holds a shared_ptr), so the original resource won't be released

Reply:
You need to be careful to use shape from the same NDArray in your NDArrayDLManager

Reply:
Thanks.

In the other case,
```python
from torch.utils import dlpack
a = torch.array([1,2,3])
pack = dlpack.to_dlpack(a)
b = mx.nd.from_dlpack(pack)
del a, pack
```
When `dlpack.to_dlpack` is called, PyTorch will allocate `ATenDLMTensor` which increases the refcount of Torch Tensor [code](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/DLConvertor.cpp#L129).
After the variables `a` and `pack` are released, `ATenDLMTensor` still exists.
I think the deleter should be called by the new NDArray `b` when the NDArray `b` releases. Refer to [PyTorch FromDLPack](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/DLConvertor.cpp#L170). However, NDArray doesn't have explicit deleter parameter.

In my PR, `from_dlpack` will copy the dlpack object.
When the old dlpack releases, it doesn't call the deleter.
The new dlpack will be a member of the new NDArray as `NDArray(handle=handle, dlpack=dlpack_copy)`.
When the new NDArray releases, the new dlpack will be released, then call the deleter by the new dlpack. And the deleter will release `NDArrayDLManager` or `ATenDLMTensor`. The refcount of the old NDArray will decrease 1.

Reply:
I made a copy of NDArray as the member of NDArrayDLManager, and the copy increase the refcount.
I'm confused how to modify the PR.
After creating a new NDArray(Python) from DLPack, then delete the old NDArray(Python) and PyCapsule(DLPack).

Which object will call the deleter function?

In my case, when `a` gets deleted, how does `b` hold the NDArrayDLManager? It seems that `b` only get the pure data pointer from `DLManagedTensor::dl_tensor::data`, the type of the pointer is not a shared pointer.
And how does `b` store the pointer to `NDArrayDLManager` in MXNet NDArray?